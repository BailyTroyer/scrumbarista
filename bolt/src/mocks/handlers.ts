import { rest } from "msw";

export const handlers = [
  rest.get("http://localhost:8000/healthy", (req, res, ctx) => {
    return res(ctx.status(200), ctx.text("healthy"));
  }),

  // rest.get("http://localhost:8000/standups", (req, res, ctx) => {
  //   return res(ctx.status(200), ctx.json({
  //     id: 'id',
  //     name: 'name',
  //     channelId: 'C01LQPT2LMD',
  //     questions: 'questions',
  //     days: ['monday']
  //   }));
  // }),

  // rest.post('https://slack.com/api/views.open', (req, res, ctx) => {
  //   return res(ctx.status(200))
  // })
];
