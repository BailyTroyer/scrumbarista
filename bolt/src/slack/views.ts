import {
  Middleware,
  SlackViewMiddlewareArgs,
  ViewSubmitAction,
} from "@slack/bolt";

import { sectionBlock } from "../blocks";
import { Day, NewStandup } from "../models";
import {
  addCheckin,
  createStandup,
  getCheckins,
  getStandup,
  updateCheckin,
  updateStandup,
} from "../services/api";

export const standupView: Middleware<
  SlackViewMiddlewareArgs<ViewSubmitAction>
> = async ({
  ack,
  view,
  client,
  body: {
    user: { id: userId },
  },
}) => {
  await ack();

  const selectedValues = view.state.values;
  const name: string = selectedValues.name.name.value;
  const channelId: string = JSON.parse(view.private_metadata).channelId;
  const startTime: string = selectedValues.time.time.selected_time;

  // check if standup already exists
  const standup = await getStandup(channelId);
  const newStandup: NewStandup = {
    name,
    channelId,
    questions: selectedValues.questions.questions.value.split("\n"),
    days: selectedValues.days.days.selected_options.map((d: any) =>
      String(d.value)
    ) as Day[],
    startTime,
    introMessage: "",
  };

  if (standup) {
    const updatedStandup = await updateStandup(channelId, newStandup);
    await client.chat.postEphemeral({
      channel: channelId,
      user: userId,
      text: updatedStandup
        ? `I've updated ${updatedStandup.name}`
        : `I was unable to update ${standup.name}`,
    });
  } else {
    const createdStandup = await createStandup(newStandup);
    await client.chat.postEphemeral({
      channel: channelId,
      user: userId,
      text: createdStandup
        ? `I've created ${createdStandup.name}`
        : `I was unable to create ${newStandup.name}`,
    });
  }
};

export const checkinView: Middleware<
  SlackViewMiddlewareArgs<ViewSubmitAction>
> = async ({
  ack,
  body: {
    user: { id: userId },
  },
  view,
  client,
}) => {
  await ack();

  // get selected form fields
  const selectedValues = view.state.values;

  const standup = await getStandup(JSON.parse(view.private_metadata).channelId);

  // if standup DNE post ephemeral message no standup
  const questions = standup.questions;
  const answers = Object.keys(selectedValues).map(
    (k) => selectedValues[k][k].value
  );
  const date = new Date().toLocaleDateString();

  const channel = String(JSON.parse(view.private_metadata).channelId);

  const user = await client.users.info({
    user: userId,
    include_locale: true,
  });

  const checkins = (await getCheckins(channel, userId, date)) || [];

  if (checkins.length === 0) {
    const updateStandupMessage = await client.chat.postMessage({
      channel,
      attachments: [
        {
          color: "#41BC88",
          blocks: [sectionBlock(`*${standup.name}*\n${date}`)],
        },
        ...questions.map((question: string, index: number) => ({
          color: "E4E4E4",
          blocks: [sectionBlock(`*${question}*\n${answers[index]}`)],
        })),
      ],
      username: user.user.profile.real_name,
      icon_url: user.user.profile.image_192,
    });

    const createCheckinDto = {
      answers,
      postMessageTs: updateStandupMessage.ts,
      userId: user.user.id,
    };

    // add checkin
    await addCheckin(channel, createCheckinDto);
  } else {
    const standupMessage = await client.chat.update({
      channel,
      attachments: [
        {
          color: "#41BC88",
          blocks: [sectionBlock(`*${standup.name}*\n${date}`)],
        },
        ...questions.map((question: string, index: number) => ({
          color: "E4E4E4",
          blocks: [sectionBlock(`*${question}*\n${answers[index]}`)],
        })),
      ],
      username: user.user.profile.real_name,
      icon_url: user.user.profile.image_192,
      ts: checkins[0].postMessageTs,
    });

    const updateCheckinDto = {
      answers,
      postMessageTs: standupMessage.ts,
    };

    // update checkin
    await updateCheckin(channel, updateCheckinDto, checkins[0].id);
  }
};
