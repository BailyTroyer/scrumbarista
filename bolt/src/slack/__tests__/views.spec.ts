import { rest } from "msw";

import { server } from "../../mocks/server";
import { checkinView, standupView } from "../views";

describe("The views handlers", () => {
  describe("standupView", () => {
    it("posts an ephemeral message when standup is successfully configured", async () => {
      const postEphemeral = jest.fn();

      server.use(
        rest.post("http://localhost:8000/standups", (req, res, ctx) =>
          res(
            ctx.json({
              id: "id",
              name: "name",
              channelId: "C01LQPT2LMD",
              questions: "questions",
              days: ["monday"],
            })
          )
        )
      );

      await standupView({
        ack: jest.fn(),
        command: {
          user_id: "user",
          channel_id: "channel",
          trigger_id: "trigger",
        },
        body: {
          user: { id: "userId" },
        },
        view: {
          state: {
            values: {
              name: { name: { value: "name" } },
              questions: { questions: { value: "questions" } },
              days: { days: { selected_options: [{ value: "day" }] } },
            },
          },
          private_metadata: JSON.stringify({ channelId: "" }),
        },
        client: {
          chat: {
            postEphemeral,
          },
        },
      } as any);

      expect(postEphemeral.mock.calls[0]).toMatchSnapshot();
    });

    it("posts an emphemeral message when unable to update standup", async () => {
      const postEphemeral = jest.fn();

      server.use(
        rest.post("http://localhost:8000/standups", (req, res, ctx) =>
          res(ctx.status(400))
        )
      );

      await standupView({
        ack: jest.fn(),
        command: {
          user_id: "user",
          channel_id: "channel",
          trigger_id: "trigger",
        },
        body: {
          user: { id: "userId" },
        },
        view: {
          state: {
            values: {
              name: { name: { value: "name" } },
              questions: { questions: { value: "questions" } },
              days: { days: { selected_options: [{ value: "day" }] } },
            },
          },
          private_metadata: JSON.stringify({ channelId: "" }),
        },
        client: {
          chat: {
            postEphemeral,
          },
        },
      } as any);

      expect(postEphemeral.mock.calls[0]).toMatchSnapshot();
    });
  });

  describe("checkinView", () => {
    it("creates new checkin and posts message in standup channel", async () => {
      const postMessage = jest.fn().mockReturnValue({
        ts: "messageTs",
      });
      const info = jest.fn().mockReturnValue({
        user: {
          id: "userId",
          profile: {
            real_name: "name",
            image_192: "image",
          },
        },
      });

      server.use(
        rest.get("http://localhost:8000/standups/channel", (req, res, ctx) =>
          res(
            ctx.json({
              id: "id",
              name: "name",
              channelId: "C01LQPT2LMD",
              questions: "questions",
              days: ["monday"],
            })
          )
        ),
        rest.post(
          "http://localhost:8000/standups/channel/checkins",
          (req, res, ctx) =>
            res(
              ctx.json({
                createdDate: new Date(),
                answers: "answers",
                postMessageTs: "messageTs",
              })
            )
        ),
        rest.get(
          "http://localhost:8000/standups/channel/checkins",
          (req, res, ctx) => res(ctx.status(400))
        )
      );

      await checkinView({
        ack: jest.fn(),
        command: {
          user_id: "user",
          channel_id: "channel",
          trigger_id: "trigger",
        },
        body: {
          user: { id: "userId" },
        },
        view: {
          state: {
            values: {
              name: { name: { value: "name" } },
              questions: { questions: { value: "questions" } },
              days: { days: { selected_options: [{ value: "day" }] } },
            },
          },
          private_metadata: JSON.stringify({ channelId: "channel" }),
        },
        client: {
          chat: {
            postMessage,
          },
          users: {
            info,
          },
        },
      } as any);

      expect(info.mock.calls[0]).toMatchSnapshot();
      expect(postMessage.mock.calls[0]).toMatchSnapshot();
    });

    it("updates pre-existing checkin and posts message in standup channel", async () => {
      const update = jest.fn().mockReturnValue({
        ts: "messageTs",
      });
      const info = jest.fn().mockReturnValue({
        user: {
          id: "userId",
          profile: {
            real_name: "name",
            image_192: "image",
          },
        },
      });

      server.use(
        rest.get("http://localhost:8000/standups/channel", (req, res, ctx) =>
          res(
            ctx.json({
              id: "id",
              name: "name",
              channelId: "C01LQPT2LMD",
              questions: "questions",
              days: ["monday"],
            })
          )
        ),
        rest.patch(
          "http://localhost:8000/standups/channel/checkins/id",
          (req, res, ctx) =>
            res(
              ctx.json({
                createdDate: new Date(),
                answers: "answers",
                postMessageTs: "messageTs",
              })
            )
        ),
        rest.get(
          "http://localhost:8000/standups/channel/checkins",
          (req, res, ctx) =>
            res(
              ctx.json([
                {
                  id: "id",
                  createdDate: new Date(),
                  answers: "answers",
                  postMessageTs: "messageTs",
                },
              ])
            )
        )
      );

      await checkinView({
        ack: jest.fn(),
        command: {
          user_id: "user",
          channel_id: "channel",
          trigger_id: "trigger",
        },
        body: {
          user: { id: "userId" },
        },
        view: {
          state: {
            values: {
              name: { name: { value: "name" } },
              questions: { questions: { value: "questions" } },
              days: { days: { selected_options: [{ value: "day" }] } },
            },
          },
          private_metadata: JSON.stringify({ channelId: "channel" }),
        },
        client: {
          chat: {
            update,
          },
          users: {
            info,
          },
        },
      } as any);

      expect(info.mock.calls[0]).toMatchSnapshot();
      expect(update.mock.calls[0]).toMatchSnapshot();
    });
  });
});
