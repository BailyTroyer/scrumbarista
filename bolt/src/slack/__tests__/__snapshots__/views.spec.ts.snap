// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The views handlers checkinView creates new checkin and posts message in standup channel 1`] = `
Array [
  Object {
    "include_locale": true,
    "user": "userId",
  },
]
`;

exports[`The views handlers checkinView creates new checkin and posts message in standup channel 2`] = `
Array [
  Object {
    "attachments": Array [
      Object {
        "blocks": Array [
          Object {
            "text": Object {
              "text": "*name*
5/9/1998",
              "type": "mrkdwn",
            },
            "type": "section",
          },
        ],
        "color": "#41BC88",
      },
      Object {
        "blocks": Array [
          Object {
            "text": Object {
              "text": "*questions*
name",
              "type": "mrkdwn",
            },
            "type": "section",
          },
        ],
        "color": "E4E4E4",
      },
    ],
    "channel": "channel",
    "icon_url": "image",
    "username": "name",
  },
]
`;

exports[`The views handlers checkinView updates pre-existing checkin and posts message in standup channel 1`] = `
Array [
  Object {
    "include_locale": true,
    "user": "userId",
  },
]
`;

exports[`The views handlers checkinView updates pre-existing checkin and posts message in standup channel 2`] = `
Array [
  Object {
    "attachments": Array [
      Object {
        "blocks": Array [
          Object {
            "text": Object {
              "text": "*name*
5/9/1998",
              "type": "mrkdwn",
            },
            "type": "section",
          },
        ],
        "color": "#41BC88",
      },
      Object {
        "blocks": Array [
          Object {
            "text": Object {
              "text": "*questions*
name",
              "type": "mrkdwn",
            },
            "type": "section",
          },
        ],
        "color": "E4E4E4",
      },
    ],
    "channel": "channel",
    "icon_url": "image",
    "ts": "messageTs",
    "username": "name",
  },
]
`;

exports[`The views handlers standupView posts an emphemeral message when unable to update standup 1`] = `
Array [
  Object {
    "channel": "",
    "text": "I was unable to create name",
    "user": "userId",
  },
]
`;

exports[`The views handlers standupView posts an ephemeral message when standup is successfully configured 1`] = `
Array [
  Object {
    "channel": "",
    "text": "I've created name",
    "user": "userId",
  },
]
`;
